### DISCUSSION TITLE
get_parent() can't return parent

### DISCUSSION THREAD

**Discussion by xydan83** [FreeOpcUa/opcua-asyncio#679] [Q&A]
Using the library, I'm trying to get nodes from KEPServerEx6. Noticed that from these nodes the parent node is not returned when trying to call get_parent(). It turns out that the KEP does not have a reverse link. 
![image](https://user-images.githubusercontent.com/86729310/136016123-3a00cf1e-c6cb-43f8-8f8d-a35c83709fcf.png)
How then can I get parent nodes?

I noticed the following entries in nodeID:
`ns=2;s=OPC Chanel.Device1.vPLC1.myNewStaticObject1.myNewStaticObject1_1.static_param1`

Is the node identifier a full path to it and can parent links be derived from it? Has anyone tried it?

**Comment by AndreasHeine** [comment]
the string name is the path in KEPServerEX6! but not on all OPC UA Servers its a implementation detail of kepware.

**Comment by oroulet** [comment]
Get parent , is the only reliable way to get parent of a node. Kepware server i broken if it does not have that information.
Some servers put the path as the string identifier. Maybe kepware does it but then this is very specific to your server

**Reply by AndreasHeine** [reply]
confirmed kepware problem...
![grafik](https://user-images.githubusercontent.com/56362817/136018852-762602bd-0e90-44bb-b20f-76524285289a.png)

**Reply by xydan83** [reply]
Kepware seems to need a separate method to get the parent of the node.

**Comment by xydan83** [comment]
I tried to write a variant of the method that can determine the ancestor by a string identifier. I'm not sure if I wrote it correctly, and for sure this method may have drawbacks, but so far this is what came to mind.

As a result, now it turns out to take parent nodes if the server is, for example, KepServerEx6.

```python
    async def get_parent(self):
        """
        returns parent of the node.
        A Node may have several parents, the first found is returned.
        This method uses reverse references, a node might be missing such a link,
        thus we will not find its parent.
        """
        refs = await self.get_references(refs=ua.ObjectIds.HierarchicalReferences, direction=ua.BrowseDirection.Inverse)
        node = None
        if len(refs) > 0:
            node = Node(self.server, refs[0].NodeId)
        else:
            if self.nodeid.NodeIdType == NodeIdType.String:
                node_id = NodeId(NamespaceIndex=self.nodeid.NamespaceIndex, Identifier='.'.join(self.nodeid.Identifier.split('.')[:-1]))
                if not len(node_id.Identifier) > 0:
                    # I'm not sure if I did the right thing, but I haven't yet found a way to determine the exact belonging of a node to any base
                    # node of the specification, since there is no reference to the parent, so by default we assume that such nodes have an 
                    # ancestor in the form of ObjectsFolder (i = 85)
                    node_id = ua.ObjectIds.ObjectsFolder
                node = Node(self.server, node_id)
        return node
```

**Reply by AndreasHeine** [reply]
why modifying the code of node-class? there is no rule that servers have to implement it that way kepware does!? its an issue on kepware side!

**Reply by xydan83** [reply]
Do I understand correctly that **KEPServer** interprets the specifications in his own way or is it just a bug? What do you think?
If so, then maybe someone else will come in handy as a workaround. At a minimum, users working with KEP can simply override the method in their Node (some `class ExtNode(Node):`, for example) inherited class, it is not necessary to modify the original, if this is really not the standard.

**Reply by AndreasHeine** [reply]
> Do I understand correctly that **KEPServer** interprets the specifications in his own way or is it just a bug? What do you think?
> If so, then maybe someone else will come in handy as a workaround. At a minimum, users working with KEP can simply override the method in their Node (some `class ExtNode(Node):`, for example) inherited class, it is not necessary to modify the original, if this is really not the standard.

i would say they dont implement it!
i am not sure if it is valid to have only forward references (it sound weird).

**Reply by xydan83** [reply]
In the spec, I did not find precise definitions of whether there should be reverse references or not ... (
